File: November2021/C++/Product_of_Array_Except_Self.cpp
// author:shyam2520

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> ans(nums.size(),1);
        for(int i=nums.size()-2;i>=0;i--)
            ans[i]=nums[i+1]*ans[i+1];
        int prefix=1;
        for(int i=0;i<nums.size();i++)
        {
            ans[i]=prefix*ans[i];
            prefix*=nums[i];
        }
        
        return ans;
    }
};

File: November2021/ProductOfArrayExceptSelf.java
class Solution {
	// TC : O(n)
	// SC : O(n)
	public int[] productExceptSelf(int[] nums) {
		int[] ans= new int[nums.length];
		int[] left = new int[nums.length];
		int[] right = new int[nums.length];
		int leftPrefixProduct = 1;
		for(int i=0;i<nums.length;i++){
			leftPrefixProduct *= nums[i];
			left[i]= leftPrefixProduct;
		}

		int rightPrefixProduct = 1;

		for(int i=nums.length-1;i>=0;i--){
			rightPrefixProduct *= nums[i] ;
			right[i]= rightPrefixProduct;
		}



		for(int i=0;i<nums.length;i++){
			int leftProduct = i>=1 ? left[i-1]:1;
			int rightProduct = i<nums.length-1 ?right[i+1]:1;
			ans[i]= leftProduct * rightProduct;
		}

		return ans;

	}
}